ntbs-service/Services/NotificationCloningService.cs

```diff
@@ -0,0 +1,109 @@
+﻿using Microsoft.EntityFrameworkCore;
+using MoreLinq.Extensions;
+using ntbs_service.DataAccess;
+using ntbs_service.Models.Entities;
+
+namespace ntbs_service.Services
+{
+    public interface INotificationCloningService
+    {
+        Notification Clone(Notification notification);
+    }
+
+    public class NotificationCloningService : INotificationCloningService
+    {
+        private readonly NtbsContext _context;
+
+        public NotificationCloningService(NtbsContext context)
+        {
+            _context = context;
+        }
+
+        public Notification Clone(Notification notification)
+        {
+            // Every primary record needs to go through the process of:
```

Comment:
yeah, haha, it was bit of a faff :P
So in this instance, if any of these were missing, the cloning failed - therefore I made sure to add an integration test that covered this.
I agree there's an overall growing number of places and I'm afraid I don't know if there's one good place to cover it all. I think it is through a combination of tests like here and writing the code in such a way that it's hard/awkward to NOT add it in. For example we used to initialise all the owned submodels in the Create route handler, we then moved it to the constructor of Notification. The latter means that
a) we can't ever create a notification that's lacking submodels and
b) when adding a new submodel, you already have to modify Notification.cs to add it to the list of navigation properties. So you're more likely to notice that a change needs to happen in the constructor, too, rather than having to remember another file

Perhaps an alternative would involve some sort of reflection-based approach, but I'm a little worried it could get quite hairy quite quickly like that